source/common/http/conn_manager_impl.cc

```diff
@@ -1135,6 +1135,73 @@ void ConnectionManagerImpl::ActiveStream::snapScopedRouteConfig() {
 
 void ConnectionManagerImpl::ActiveStream::refreshCachedRoute() { refreshCachedRoute(nullptr); }
 
+void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
+  if (!filter_manager_.streamInfo().route_entry_ || !request_headers_) {
+    return;
+  }
+  auto& route = filter_manager_.streamInfo().route_entry_;
+
+  auto grpc_timeout = Grpc::Common::getGrpcTimeout(*request_headers_);
+  std::chrono::milliseconds timeout = std::chrono::milliseconds(0);
+
+  if (!grpc_timeout || !route->grpcTimeoutHeaderMax()) {
+    // Either there is no grpc-timeout header or special timeouts for it are not
+    // configured. Use stream duration.
+    if (route->maxStreamDuration()) {
+      timeout = route->maxStreamDuration().value();
+    } else {
+      // Fall back to HCM config.
+      const auto max_stream_duration = connection_manager_.config_.maxStreamDuration();
+      if (max_stream_duration.has_value() && max_stream_duration.value().count()) {
+        timeout = max_stream_duration.value();
+      }
+    }
+  } else {
+    // Start with the timeout equal to the gRPC timeout header.
+    timeout = grpc_timeout.value();
+    // If there's a valid cap, apply it.
+    if ((timeout > route->grpcTimeoutHeaderMax().value() ||
+         timeout == std::chrono::milliseconds(0)) &&
```

Comment:
Why special-case timeout==0?  I don't think that should ever happen on the wire, but if it did, it would probably actually mean that the RPC should be cancelled immediately.